### PR TITLE
[Ruff 0.6] Stabilise 9 pylint rules

### PR DESCRIPTION
**Pull request by AlexWaygood**
## Summary

Stabilise the following rules for Ruff 0.6:
- `singledispatch-method` (`PLE1519`)
- `singledispatchmethod-function` (`PLE1520`)
- `bad-staticmethod-argument` (`PLW0211`)
- `if-stmt-min-max` (`PLR1730`)
- `invalid-bytes-return-type` (`PLE0308`)
- `invalid-hash-return-type` (`PLE0309`)
- `invalid-index-return-type` (`PLE0305`)
- `invalid-length-return-type` (`E303`)
- ~~`non-augmented-assignment` (`PLR6104`)~~
- `self-or-cls-assignment` (`PLW0642`)

These rules all seem fairly uncontroversial, they've all been in preview for >3 months (and haven't changed much in a while), and none of them have any significant issues open about them that should block stabilisation IMO.

## Test Plan

`cargo test`, but the ecoystem report will be the real judge...

### CODE REVIEW

**Comment by AlexWaygood** [comment]
There's a lot of `PLW0642` hits on `pandas`, but I think they're all true positives. `pandas` apparently does a lot of `self = foo` assignments, but I think these could easily be replaced with other variable names, which would improve the readability of the code.

**Review by MichaReiser — APPROVED** [review]

**Comment by AlexWaygood** [comment]
> LGTM. My only feedback for `PLW0642` is that I find the message: _Invalid assignment to_ somewhat confusing. I'm not familiar enough with Python but the assignment itself isn't _invalid_. It's just not recommended to change the `cls` or `self` variable but I don't think it breaks anything, it's just unexpected.

Yeah, that's a great point. I think I'll change it to "confusing". It's not an _invalid_ assignment; you could just write the code in a clearer/less confusing way by using a different variable name.

File: crates/ruff_linter/src/rules/pylint/rules/self_or_cls_assignment.rs [hunk 1]
```diff
@@ -47,10 +53,14 @@ impl Violation for SelfOrClsAssignment {
...
         format!(
-            "Invalid assignment to `{}` argument in {method_type} method",
+            "Confusing assignment to `{}` argument in {method_type} method",
             method_type.arg_name(),
         )
     }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Consider using a different variable name".to_string())
+    }
```
Comment (dhruvmanila) [inline crates/ruff_linter/src/rules/pylint/rules/self_or_cls_assignment.rs:63]: This is clever ;)

File: crates/ruff_linter/src/rules/pylint/rules/self_or_cls_assignment.rs [hunk 2]
```diff
@@ -47,10 +53,14 @@ impl Violation for SelfOrClsAssignment {
         let SelfOrClsAssignment { method_type } = self;
 
         format!(
-            "Invalid assignment to `{}` argument in {method_type} method",
+            "Confusing assignment to `{}` argument in {method_type} method",
             method_type.arg_name(),
```
Comment (dhruvmanila) [inline crates/ruff_linter/src/rules/pylint/rules/self_or_cls_assignment.rs:57]: Another option could be "Re-assigned `{}` argument in {method_type} method"

Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/pylint/rules/self_or_cls_assignment.rs:57]: Oh, that's better. Much more descriptive. I'll make another PR 😄

**Pull request merged**